### PR TITLE
Generate: move `logits` to same device as `input_ids`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

In `generate`, we create many auxiliary tensors for internal operations. Prior to this commit, the assumption was that the model's output (`logits`) and these auxiliary tensors were on the same device. In practice, this is the same as assuming that `input_ids` (or other main input, depending on the model) would be on the same device as `lm_head`.

This assumption doesn't always hold, even if the user follows our suggested examples. For instance, if we do `model_inputs = model_inputs.to(model.device)`, as we suggest in many examples, we move `model_inputs` to the same device as in the first set of parameters in `model.parameters()`, which is far from guaranteed to be the same device as `lm_head`. ([PreTrainedModel.device source code](https://github.com/huggingface/transformers/blob/24b82f3cd56d5eeb26e649207aac3ce8a7d75bdc/src/transformers/modeling_utils.py#L1072), [.parameters() docs](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.parameters))

To make the assumption hold, this PR moves the `logits` returned by the model to the same device as `input_ids`, which should fix most (if not all) of device issues in multi-device settings.

Example of command fixed by this PR: `CUDA_VISIBLE_DEVICES=0,1 RUN_SLOW=1 python -m pytest tests/quantization/bnb/test_mixed_int8.py -k test_generate_quality_config -vv` ✅ 

_________________________________

### ⚠️ caveats

The following script now runs. 

```py
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("distilgpt2")
model = AutoModelForCausalLM.from_pretrained("distilgpt2", device_map="auto")

inputs = tokenizer(["The quick brown"], return_tensors="pt")
gen_out = model.generate(**inputs, do_sample=False)
```

Note that `input_ids` is on the CPU, so some internal ops in `generate` will run on the CPU (while the forward pass happens on GPU). We throw a [loud warning](https://github.com/huggingface/transformers/blob/24b82f3cd56d5eeb26e649207aac3ce8a7d75bdc/src/transformers/generation/utils.py#L2090), but this may still be a source of issues (`generate` running slower than expected because there are CPU ops and device transfers)

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
Thanks for fixing this! LGTM!

**Review by ArthurZucker — APPROVED** [review]
👁️ 👁️ this should help with all cache issues I think nice!
